x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    void perform(Runnable onComplete) {
+        job = null;
+        clusterService.submitStateUpdateTask("update_tsdb_data_stream_end_times", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return updateTimeSeriesTemporalRange(currentState, Instant.now());
+            }
+
+            @Override
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                onComplete.run();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                LOGGER.warn("failed to update tsdb data stream end times", e);
+                onComplete.run();
+            }
+
+        }, ClusterStateTaskExecutor.unbatched());
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating ["
+                + DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey()
+                + "] setting from ["
+                + pollInterval
+                + "] to ["
+                + newValue
+                + "]"
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            String head = dataStream.getWriteIndex().getName();
+            IndexMetadata im = current.metadata().getIndices().get(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                Settings settings = Settings.builder()
+                    .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                    .build();
+                LOGGER.debug(
+                    "updating [{}] setting from [{}] to [{}] for index [{}]",
+                    IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                    currentEnd,
+                    newEnd,
+                    head
+                );
+                if (mBuilder == null) {
+                    mBuilder = Metadata.builder(current.metadata());
+                }
+                mBuilder.updateSettings(settings, head);
+                Metadata.Builder finalMBuilder = mBuilder;
```

Comment:
This line is a little strange here, why this instead of passing `mBuilder::get` below?

Reply:
Yes, this is a little strange, but `mBuilder` is because is initially initialised as null and only if there is a need to update end_time setting then a new builder is initialised. So we can't pass down `mBuilder` here, unless we always initialise the builder (at the cost of initialising a builder based on an existing `Metadata` instance).